omnipaxos/src/storage.rs

```diff
@@ -796,12 +696,56 @@ where
         self.state_cache.stopsign_is_decided()
     }
 
-    pub(crate) fn create_snapshot(&mut self, compact_idx: u64) -> StorageResult<T::Snapshot> {
-        let compacted_idx = self.get_compacted_idx();
-        if compact_idx < compacted_idx {
-            Err(CompactionErr::TrimmedIndex(compacted_idx))?
+    pub(crate) fn sync_log(
+        &mut self,
+        accepted_round: Ballot,
+        decided_idx: usize,
+        log_sync: Option<LogSync<T>>,
+    ) -> StorageResult<usize> {
+        self.state_cache.accepted_round = accepted_round;
+        self.state_cache.decided_idx = decided_idx;
+        let mut sync_txn: Vec<StorageOp<T>> = vec![
+            StorageOp::SetAcceptedRound(accepted_round),
+            StorageOp::SetDecidedIndex(decided_idx),
+        ];
+        if let Some(sync) = log_sync {
+            match sync.decided_snapshot {
+                Some(SnapshotType::Complete(c)) => {
+                    self.state_cache.compacted_idx = sync.sync_idx;
+                    sync_txn.push(StorageOp::Trim(sync.sync_idx));
+                    sync_txn.push(StorageOp::SetCompactedIdx(sync.sync_idx));
+                    sync_txn.push(StorageOp::SetSnapshot(Some(c)));
+                }
+                Some(SnapshotType::Delta(d)) => {
+                    let mut snapshot = self.create_decided_snapshot()?;
+                    snapshot.merge(d);
+                    self.state_cache.compacted_idx = sync.sync_idx;
+                    sync_txn.push(StorageOp::Trim(sync.sync_idx));
+                    sync_txn.push(StorageOp::SetCompactedIdx(sync.sync_idx));
+                    sync_txn.push(StorageOp::SetSnapshot(Some(snapshot)));
+                }
+                None => (),
+            }
+            self.state_cache.accepted_idx = sync.sync_idx + sync.suffix.len();
+            sync_txn.push(StorageOp::AppendOnPrefix(sync.sync_idx, sync.suffix));
+            self.state_cache.stopsign = sync.stopsign.clone();
+            if sync.stopsign.is_some() {
+                self.state_cache.accepted_idx += 1;
+            }
+            sync_txn.push(StorageOp::SetStopsign(sync.stopsign));
```

Comment:
setting the stopsign must be inside the if?

Reply:
Because of the split-brain case, a LogSync with stopsign = None actually means overwrite whatever stopsign you have with None. The LogSync struct is kinda confusing because some fields are diffs between the logs (snapshots when delta and suffix) but some are absolute (decided_idx, accepted_round, stopsign) where you just adopt them even if they are the same as what you already have. Do you think this needs redesigning?

Reply:
I see. I think it's ok for now. But here we can just make this logic more explicit i.e., still do the `match` on the StopSign and in the `None` case we set the stopsign to None instead.